### PR TITLE
bind_cols handles unnamed list. closes #3402

### LINKED ISSUE #3402: bind_cols() gives cryptic error message for unnamed inputs
``` r
dplyr::bind_cols(list(1, 2))
#> Error in cbind_all(x): Not compatible with STRSXP: [type=NULL].
dplyr::bind_cols(list(one = 1, two = 2))
#> # A tibble: 1 x 2
#>     one   two
#>   <dbl> <dbl>
#> 1    1.    2.
```

Created on 2018-03-09 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

Possibly relevant context: I came across this via `purrr::map_dfc()`. So the fact you need names and how/where to inject them is less obvious than the above reprex indicates. Hence a better error message would really pay off.

### PR DESCRIPTION
**Pull request by romainfrancois**
- bind_cols handles unnamed list. (#3402)

--- 

I made it so that 
 
```
> bind_cols(list(1, 2))
# A tibble: 1 x 2
     V1    V2
  <dbl> <dbl>
1    1.    2.
> bind_cols(list(1, x=2))
# A tibble: 1 x 2
     V1     x
  <dbl> <dbl>
1    1.    2.
> bind_cols(list(x=1, 2))
# A tibble: 1 x 2
      x    V1
  <dbl> <dbl>
1    1.    2.
```

but if it's better I can make it fail with a better error message.

### CODE REVIEW

**Comment by krlmlr** [comment]
Do we really need to support this operation for unnamed lists?

Last time, the "closes #xxxx" entry in the issue title wasn't picked up by GitHub. It works from the body of the PR, though.

**Comment by romainfrancois** [comment]
It works for partially named lists already.

The `closes` in the PR title comes from the commit message of the one commit. I don't understand why this would not work, unless you squash perhaps ?

File: src/utils.cpp [hunk 1]
```diff
@@ -277,6 +277,14 @@ SEXP vec_names(SEXP x) {
   return Rf_getAttrib(x, R_NamesSymbol);
 }
 
+SEXP vec_names_or_empty(SEXP x) {
+  SEXP nms = Rf_getAttrib(x, R_NamesSymbol);
+  if (Rf_isNull(nms)) {
+    return Rf_allocVector(STRSXP, LENGTH(x)) ;
```
Comment (krlmlr) [inline src/utils.cpp]: Can we agree to remove the space before the semicolon? Unfortunately, `astyle` doesn't seem to have a rule for it, but maybe `clang-format` does?

File: tests/testthat/test-binds.R [hunk 2]
```diff
@@ -622,3 +622,9 @@ test_that("unnamed vectors fail", {
 test_that("supports NULL values", {
   expect_identical(bind_cols(a = 1, NULL, b = 2, NULL), tibble(a = 1, b = 2))
 })
+
+test_that("bind_cols handles unnamed list (#3402)", {
+  expect_identical(
+    bind_cols(list(1, 2))
```
Comment (krlmlr) [inline tests/testthat/test-binds.R]: Identical to what? ;-)

File: src/utils.cpp [hunk 3]
```diff
@@ -277,6 +277,14 @@ SEXP vec_names(SEXP x) {
   return Rf_getAttrib(x, R_NamesSymbol);
 }
 
+SEXP vec_names_or_empty(SEXP x) {
```
Comment (krlmlr) [inline src/utils.cpp]: @lionel-: Can we access `names2()` from *rlang*?

Comment (lionel-) [reply src/utils.cpp]: There is no exported C interface for names2.

File: tests/testthat/test-binds.R (hunk 2 above)
Comment (romainfrancois) [reply tests/testthat/test-binds.R]: oups, fixed now.

**Comment by jennybc** [comment]
This is an occasion to revisit conversations about a consistent approach for manufacturing (or de-duplicating) names (https://github.com/tidyverse/tibble/issues/217, https://github.com/tidyverse/tibble/pull/248).

@krlmlr My impression is that `tibble::tidy_names()` is the preferred way to handle this now. `bind_cols()` [currently calls `tibble::repair_names()`](https://github.com/tidyverse/dplyr/blob/0818cab40c109b88777b7c4e9e6b768dcca4e8c0/R/bind.r#L139). However, perhaps this would be too disruptive and/or does not belong in this PR.

**Comment by romainfrancois** [comment]
Perhaps this is a different issue indeed. I just wanted to double check if the syntax with the unnamed list should fail (with a non cryptic msg) or come up with names. 

I guess this is good for now.

**Pull request merged**